File: golang/shm2u.go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"strconv"
	"github.com/hslam/mmap"
	"github.com/hslam/shm"
)

func helper(cmd string) {
	fmt.Printf("example: %v write msg\n", cmd)
	fmt.Printf("         %v read size\n", cmd)
}

func main() {
	name := "test"
	size := 128

	if len(os.Args) < 3 {
		helper(os.Args[0])
		return
	}

	if os.Args[1] == "read" {
		sz, _ := strconv.Atoi(os.Args[2])

		fd, err := shm.Open(name, shm.O_RDONLY, 0644)
		if err != nil {
			panic(err)
		}
		defer shm.Close(fd)
	
		data, err := mmap.Open(fd, 0, size, mmap.READ)
		if err != nil {
			panic(err)
		}
		defer mmap.Munmap(data)
	
		fmt.Printf("%p:%s\n", data, string(data[:sz]))
	} else if os.Args[1] == "write" {
		msg := os.Args[2]

		fd, err := shm.Open(name, shm.O_RDWR|shm.O_CREATE, 0644)
		if err != nil {
			panic(err)
		}
		defer shm.Unlink(name)
		defer shm.Close(fd)

		shm.Ftruncate(fd, int64(size))
		data, err := mmap.Open(fd, 0, size, mmap.READ|mmap.WRITE)
		if err != nil {
			panic(err)
		}
		defer mmap.Munmap(data)

		context := []byte(msg)
		copy(data, context)
		fmt.Printf("address %p:%s\n", data, msg)


		c := make(chan os.Signal, 1)
		signal.Notify(c, os.Interrupt)
		// Block until a signal is received.
		for {
			<-c
			break
		}
	} else {
		helper(os.Args[0])
		return
	}
}

File: c/shm2u.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>

static int running = 1;
void
inthandler(int sig)
{
	running = 0;
}

int
nsh_open_shmem(char *name)
{
	int fd = shm_open(name, O_CREAT | O_RDWR, 0644);
	if (fd == -1) {
		printf("open %d failed errno%d\n", fd, errno);
		return -1;
	}

	return fd;
}

int
nsh_close_shmem(char *name)
{
	int rv = shm_unlink(name);
	if (rv != 0) {
		printf("close shmem %s failed errno%d\n", name, errno);
		return -1;
	}

	return 0;
}

int
nsh_grow_shmem(int fd, size_t size)
{
	if (ftruncate(fd, size) == -1) {
		printf("ftruncate %d failed errno%d\n", fd, errno);
		return -1;
	}

	return 0;
}

void *
nsh_open_mmap(int fd, size_t size)
{
	int protection = PROT_READ | PROT_WRITE;

	// The buffer will be shared (meaning other processes can access it),
	// but anonymous (meaning third-party processes cannot obtain an
	// address for it), so only this process and its children will be able
	// to use it:
	// int visibility = MAP_SHARED | MAP_ANONYMOUS;
	int visibility = MAP_SHARED;

	return mmap(NULL, size, protection, visibility, fd, 0);
}

int
nsh_close_mmap(void *mem, size_t size)
{
	return munmap(mem, size);
}

void
helper(char *cmd)
{
	printf("example: %s write msg\n", cmd);
	printf("         %s read size\n", cmd);
}

int
main(int argc, char *argv[])
{
	char *bin = argv[0];
	if (argc < 3) {
		helper(bin);
		return 0;
	}
	char *cmd = argv[1];
	size_t size = 128;
	char *name = "test";

	if (0 == strcmp(cmd, "write")) {
		signal(SIGINT, inthandler);
		char * msg = argv[2];

		int fd = nsh_open_shmem(name);
		if (fd == -1) {
			printf("failed to open shmem errno%d\n", errno);
			return 0;
		}

		int rv = nsh_grow_shmem(fd, size);
		if (rv != 0) {
			printf("failed to grow shmem errno%d\n", errno);
			return 0;
		}

		void *shmem = nsh_open_mmap(fd, size);
		if (shmem == MAP_FAILED) {
			printf("failed to get mmap errno%d\n", errno);
			return 0;
		}

		memcpy(shmem, msg, strlen(msg) + 1);
		printf("address %p:%s\n", shmem, msg);
		while (running) {
			sleep(1);
		}

		rv = nsh_close_mmap(shmem, size);
		if (rv != 0) {
			printf("failed to close mmap errno%d\n", errno);
			return 0;
		}

		rv = nsh_close_shmem(name);
		if (rv != 0) {
			printf("failed to unlink shmem%d errno%d\n", rv, errno);
			return 0;
		}
	} else if (0 == strcmp(cmd, "read")) {
		int size = atoi(argv[2]);

		int fd = nsh_open_shmem(name);
		if (fd == -1) {
			printf("failed to open shmem errno%d\n", errno);
			return 0;
		}

		void *shmem = nsh_open_mmap(fd, size);
		if (shmem == MAP_FAILED) {
			printf("failed to get mmap errno%d\n", errno);
			return 0;
		}

		printf("%p:%.*s\n", shmem, size, (char *)shmem);
	} else {
		helper(bin);
	}
}
